File: Java/NumberOfSubarraysWithBoundedMaximum.java
public class NumberOfSubarraysWithBoundedMaximum {
    public int numSubarrayBoundedMax(int[] nums, int left, int right) {
        return count(nums, right) - count(nums, left - 1);
    }

    private int count(int[] nums, int bound) {
        int total = 0, length = 0;
        for (int x : nums) {
            length = (x <= bound) ? length + 1 : 0;
            total += length;
        }
        return total;
    }

    public static void main(String[] args) {
        NumberOfSubarraysWithBoundedMaximum s = new NumberOfSubarraysWithBoundedMaximum();
        System.out.println(s.numSubarrayBoundedMax(new int[] { 2, 1, 4, 3 }, 2, 3));    // 3
        System.out.println(s.numSubarrayBoundedMax(new int[] { 2, 9, 2, 5, 6 }, 2, 8)); // 7
        System.out.println(s.numSubarrayBoundedMax(new int[] { 1, 1, 1 }, 1, 1));       // 6
    }
}

File: Go/NumberOfSubarraysWithBoundedMaximum.go
//go:build ignore

package main

import "fmt"

func numSubarrayBoundedMax(nums []int, left, right int) int {
	count := func(bound int) int {
		total, length := 0, 0
		for _, x := range nums {
			if x <= bound {
				length++
			} else {
				length = 0
			}
			total += length
		}
		return total
	}
	return count(right) - count(left-1)
}

func main() {
	fmt.Println(numSubarrayBoundedMax([]int{2, 1, 4, 3}, 2, 3))    // 3
	fmt.Println(numSubarrayBoundedMax([]int{2, 9, 2, 5, 6}, 2, 8)) // 7
	fmt.Println(numSubarrayBoundedMax([]int{1, 1, 1}, 1, 1))       // 6
}
